src/CodeRunner.ts

```diff
@@ -0,0 +1,277 @@
+     * @param {RenderOptions} statusPanelOptions Options for rendering the panel
+     * @param {RenderOptions} inputOptions Options for rendering the InputManager
+     * @param {RenderOptions} codeEditorOptions Options for rendering the editor
+     * @return {HTMLElement} The rendered RunStateManager
+     */
+    render(statusPanelOptions: RenderOptions,
+        inputOptions: RenderOptions,
+        codeEditorOptions: RenderOptions): HTMLElement {
+        const rendered = renderWithOptions(statusPanelOptions, `
+<div class="grid grid-cols-2 items-center">
+    <div class="col-span-1 flex flex-row">
+        ${this.buttons.map(b => b.buttonHTML).join("\n")}
+    </div>
+    <div class="col-span-1 flex flex-row-reverse">
+        <div id="${APPLICATION_STATE_TEXT_ID}"></div>
+        ${svgCircle(STATE_SPINNER_ID, "red")}
+    </div>
+</div>`);
+        // Buttons are freshly added to the DOM, so attach listeners now
+        this.buttons.forEach(b => addListener(b.id, b.onClick, "click"));
+        this.inputManager.render(inputOptions);
+        this.editor.render(codeEditorOptions, rendered);
+        return rendered;
+    }
+
+    /**
+     * Run the code that is currently present in the editor
+     * @return {Promise<void>} Promise of running the code
+     */
+    async runCode(): Promise<void> {
+        // Setup pre-run
+        this.setState(RunState.Running);
+        BackendManager.publish({
+            type: BackendEventType.Start,
+            data: "User started run", contentType: "text/plain"
+        });
+        papyrosLog(LogType.Debug, "Running code in Papyros, sending to backend");
+        const start = new Date().getTime();
+        let endMessage = "Program finishd normally";
+        let terminated = false;
+        try {
+            await (this.backend).then(b => b.call(
+                b.workerProxy.runCode, this.editor.getCode()
+            ));
+        } catch (error: any) {
+            papyrosLog(LogType.Debug, "Error during code run", error);
+            if (error.type === "InterruptError") {
+                // Error signaling forceful interrupt
+                terminated = true;
+            }
+            BackendManager.publish({
+                type: BackendEventType.Error,
+                data: JSON.stringify(error),
+                contentType: "text/json"
+            });
+            endMessage = "Program terminated due to error: " + error.constructor.name;
+        } finally {
+            if (this.state !== RunState.Stopping) { // Was interrupted
+                BackendManager.publish({
+                    type: BackendEventType.End,
+                    data: endMessage, contentType: "text/plain"
+                });
+            }
+            const end = new Date().getTime();
+            this.setState(RunState.Ready, t("Papyros.finished", { time: (end - start) / 1000 }));
+            if (terminated) {
+                await this.start();
```

Comment:
comsync already restarts the worker after terminating. Also, you can end up here if you interrupt/terminate the client for the sake of switching programming language.

Reply:
Restarting the worker is not enough, as the launch method must also be called. This is done in start, so that should be fine. For the second case, I don't think anything will go wrong if start is called twice.